### PR TITLE
PostgreSQL protocol support UUID binary data format

### LINKED ISSUE #24043: java.lang.IllegalArgumentException: Cannot find PostgreSQL type 'POSTGRESQL_TYPE_UUID'
## Bug Report

**For English only**, other languages will not accept.

Before report a bug, make sure you have:

- Searched open and closed [GitHub issues](https://github.com/apache/shardingsphere/issues).
- Read documentation: [ShardingSphere Doc](https://shardingsphere.apache.org/document/current/en/overview).

Please pay attention on issues you submitted, because we maybe need more details. 
If no response anymore and we cannot reproduce it on current information, we will **close it**.

Please answer these questions before submitting your issue. Thanks!

### Which version of ShardingSphere did you use?
ShardingSphere Proxy 5.3.2

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
ShardingSphere-Proxy

### Expected behavior
Support POSTGRESQL_TYPE_UUID column type.
### Actual behavior
`java.lang.IllegalArgumentException: Cannot find PostgreSQL type 'POSTGRESQL_TYPE_UUID' in column type when process binary protocol value`
### Reason analyze (If you can)
```
java.lang.IllegalArgumentException: Cannot find PostgreSQL type 'POSTGRESQL_TYPE_UUID' in column type when process binary protocol value
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:217)
	at org.apache.shardingsphere.db.protocol.postgresql.packet.command.query.extended.bind.protocol.PostgreSQLBinaryProtocolValueFactory.getBinaryProtocolValue(PostgreSQLBinaryProtocolValueFactory.java:145)
	at org.apache.shardingsphere.db.protocol.postgresql.packet.command.query.extended.bind.PostgreSQLComBindPacket.getBinaryParameters(PostgreSQLComBindPacket.java:143)
	at org.apache.shardingsphere.db.protocol.postgresql.packet.command.query.extended.bind.PostgreSQLComBindPacket.readParameters(PostgreSQLComBindPacket.java:82)
	at org.apache.shardingsphere.proxy.frontend.postgresql.command.query.extended.bind.PostgreSQLComBindExecutor.execute(PostgreSQLComBindExecutor.java:51)
	at org.apache.shardingsphere.proxy.frontend.postgresql.command.query.extended.PostgreSQLAggregatedCommandExecutor.execute(PostgreSQLAggregatedCommandExecutor.java:41)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.executeCommand(CommandExecutorTask.java:111)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.run(CommandExecutorTask.java:78)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
```

### PR DESCRIPTION
**Pull request by jyotisharma7**
Fixes #24043.

Changes proposed in this pull request:
  - Support for UUID Datatype in ShardingSphere

---

Before committing this PR, I'm sure that I have checked the following options:
- [X] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [X] I have self-reviewed the commit code.
- [X] I have (or in comment I request) added corresponding labels for the pull request.
- [X] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [X] I have made corresponding changes to the documentation.
- [X] I have added corresponding unit tests for my changes.

### CODE REVIEW

File: db-protocol/postgresql/src/main/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/bind/protocol/PostgreSQLUUIDBinaryProtocolValue.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import org.apache.shardingsphere.db.protocol.postgresql.payload.PostgreSQLPacketPayload;
+
+/**
+ * Binary protocol value for UUID for PostgreSQL.
+ */
+public final class PostgreSQLUUIDBinaryProtocolValue implements PostgreSQLBinaryProtocolValue {
+    
+    @Override
+    public int getColumnLength(final Object value) {
+        return (value.toString()).length();
```
Comment (TeslaCN) [inline db-protocol/postgresql/src/main/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/bind/protocol/PostgreSQLUUIDBinaryProtocolValue.java:30]: Is the parentheses around `value.toString()` necessary?

File: db-protocol/postgresql/src/main/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/bind/protocol/PostgreSQLUUIDBinaryProtocolValue.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    public Object read(final PostgreSQLPacketPayload payload, final int parameterValueLength) {
+        byte[] result = new byte[parameterValueLength];
+        UUID uuidFromBytes = UUID.nameUUIDFromBytes(result);
+        payload.getByteBuf().readBytes(result);
+        return uuidFromBytes;
+    }
+    
+    @Override
+    public void write(final PostgreSQLPacketPayload payload, final Object value) {
+        payload.writeBytes((byte[]) value);
```
Comment (TeslaCN) [inline db-protocol/postgresql/src/main/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/bind/protocol/PostgreSQLUUIDBinaryProtocolValue.java:43]: Is it possible that the value is an instance of `java.util.UUID`?

File: db-protocol/postgresql/src/main/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/bind/protocol/PostgreSQLUUIDBinaryProtocolValue.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    public int getColumnLength(final Object value) {
+        return (value.toString()).length();
+    }
+    
+    @Override
+    public Object read(final PostgreSQLPacketPayload payload, final int parameterValueLength) {
+        byte[] result = new byte[parameterValueLength];
+        UUID uuidFromBytes = UUID.nameUUIDFromBytes(result);
+        payload.getByteBuf().readBytes(result);
+        return uuidFromBytes;
```
Comment (TeslaCN) [inline db-protocol/postgresql/src/main/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/bind/protocol/PostgreSQLUUIDBinaryProtocolValue.java:38]: The variable to be returned should be named `result`.

**Review by TeslaCN — CHANGES_REQUESTED** [review]

**Comment by TeslaCN** [comment]
It seems that the UUID in binary format is 16 bytes.

https://github.com/pgjdbc/pgjdbc/blob/99d010147b3e6907a39485ec0113c56d09fdfc37/pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java#L3575-L3590

File: db-protocol/postgresql/src/main/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/bind/protocol/PostgreSQLUUIDBinaryProtocolValue.java (hunk 1 above)
Comment (jyotisharma7) [reply db-protocol/postgresql/src/main/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/bind/protocol/PostgreSQLUUIDBinaryProtocolValue.java:30]: yes it is required as it is a function.

File: db-protocol/postgresql/src/main/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/bind/protocol/PostgreSQLUUIDBinaryProtocolValue.java (hunk 3 above)
Comment (jyotisharma7) [reply db-protocol/postgresql/src/main/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/bind/protocol/PostgreSQLUUIDBinaryProtocolValue.java]: I fixed this issue.

File: db-protocol/postgresql/src/main/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/bind/protocol/PostgreSQLUUIDBinaryProtocolValue.java (hunk 1 above)
Comment (TeslaCN) [reply db-protocol/postgresql/src/main/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/bind/protocol/PostgreSQLUUIDBinaryProtocolValue.java:30]: I mean:
```suggestion
        return value.toString().length();
```

**Comment by jyotisharma7** [comment]
> It seems that the UUID in binary format is 16 bytes.
> 
> https://github.com/pgjdbc/pgjdbc/blob/99d010147b3e6907a39485ec0113c56d09fdfc37/pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java#L3575-L3590



> It seems that the UUID in binary format is 16 bytes.
> 
> https://github.com/pgjdbc/pgjdbc/blob/99d010147b3e6907a39485ec0113c56d09fdfc37/pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java#L3575-L3590

yes it is 16 bytes in binary format.

File: db-protocol/postgresql/src/test/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/bind/protocol/PostgreSQLUUIDBinaryProtocolValueTest.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import java.util.UUID;
+
+import java.nio.charset.StandardCharsets;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+public final class PostgreSQLUUIDBinaryProtocolValueTest {
+    
+    private final byte[] expected = new byte[5];
```
Comment (TeslaCN) [inline db-protocol/postgresql/src/test/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/bind/protocol/PostgreSQLUUIDBinaryProtocolValueTest.java:33]: I wonder why the 5 bytes worked in unit test if UUID is 16 bytes?

**Comment by TeslaCN** [comment]
Hi @jyotisharma7 
Any update here?

**Comment by jyotisharma7** [comment]
> Hi @jyotisharma7 Any update here?

Yes, I will update the PR will latest changes.

File: db-protocol/postgresql/src/main/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/bind/protocol/PostgreSQLUUIDBinaryProtocolValue.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    
+    @Override
+    public int getColumnLength(final Object value) {
+        return value.toString().length();
+    }
+    
+    @Override
+    public Object read(final PostgreSQLPacketPayload payload, final int parameterValueLength) {
+        byte[] uuidFromBytes = new byte[parameterValueLength];
+        UUID result = UUID.nameUUIDFromBytes(uuidFromBytes);
```
Comment (TeslaCN) [inline db-protocol/postgresql/src/main/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/bind/protocol/PostgreSQLUUIDBinaryProtocolValue.java:36]: `UUID.nameUUIDFromBytes` is not for deserializing.

This is how PostgreSQL JDBC deserialize an UUID instance.
https://github.com/pgjdbc/pgjdbc/blob/99d010147b3e6907a39485ec0113c56d09fdfc37/pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java#L3589

**Pull request closed without merging**